infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftMachine.java

```diff
@@ -86,65 +88,29 @@ public String getName() {
-            for (ContainerPort containerPort : cointaner.getPorts()) {
-                Integer port = containerPort.getContainerPort();
-
-                for (Service service : matchedServices) {
-                    for (ServicePort servicePort : service.getSpec().getPorts()) {
-                        if (port.equals(servicePort.getPort())) {
-                            // container is accessible via this service
-                            matchedServicesPorts.put(service.getMetadata().getName(), servicePort);
-                        }
-                    }
-                }
-            }
+            Set<String> matchedServices = getMatchedServices().stream()
+                                                              .map(s -> s.getMetadata().getName())
+                                                              .collect(Collectors.toSet());
 
             List<Route> routes = client.routes()
                                        .inNamespace(projectName)
                                        .list()
                                        .getItems();
+
             Map<String, ServerImpl> servers = new HashMap<>();
             for (Route route : routes) {
-                String serviceName = route.getSpec().getTo().getName();
-                List<ServicePort> servicePorts = matchedServicesPorts.get(serviceName);
-                if (servicePorts != null) {
-                    RoutePort routePort = route.getSpec().getPort();
-                    for (ServicePort servicePort : servicePorts) {
-                        String portReference = routePort.getTargetPort().getStrVal();
-                        if (portReference != null) {
-                            if (portReference.equals(servicePort.getName())) {
-                                servers.put(portReference, new ServerImpl("http://" + route.getSpec().getHost(),
-                                                                          ServerStatus.UNKNOWN));
-                            }
-                            continue;
-                        }
-
-                        Integer portNumber = routePort.getTargetPort().getIntVal();
-                        if (portNumber != null) {
-                            if (portNumber.equals(servicePort.getPort())) {
-                                servers.put(Integer.toString(portNumber),
-                                            new ServerImpl("http://" + route.getSpec().getHost(),
-                                                           ServerStatus.UNKNOWN));
-                            }
-                            continue;
-                        }
+                if (matchedServices.contains(route.getSpec().getTo().getName())) {
+                    Map<String, String> annotations = route.getMetadata().getAnnotations();
+                    String serverName = annotations.get(CHE_SERVER_NAME_ANNOTATION);
```

Comment:
Maybe it makes sense to create a component similar to `org.eclipse.che.workspace.infrastructure.docker.Labels`, so you can define the constants there + implement annotations serialization and deserialization + it may be useful for feature recovery.